tf_quant_finance/black_scholes/approximations/american_option.py

```diff
@@ -343,6 +343,258 @@ def _calc_q(n, m, sign, k=1):
+    Bjerksund Stensland price of the options using the 1993 flat boundary model.
+
+  Raises:
+    ValueError: If both `forwards` and `spots` are supplied or if neither is
+      supplied.
+    ValueError: If both `discount_rates` and `discount_factors` is supplied.
+    ValueError: If both `dividend_rates` and `cost_of_carries` is
+      supplied.
+  """
+  if (spots is None) == (forwards is None):
+    raise ValueError('Either spots or forwards must be supplied but not both.')
+  if (discount_rates is not None) and (discount_factors is not None):
+    raise ValueError('At most one of discount_rates and discount_factors may '
+                     'be supplied')
+  if (dividend_rates is not None) and (cost_of_carries is not None):
+    raise ValueError('At most one of dividend_rates and cost_of_carries '
+                     'may be supplied')
+
+  with tf.name_scope(name or 'option_price'):
+    strikes = tf.convert_to_tensor(strikes, dtype=dtype, name='strikes')
+    dtype = strikes.dtype
+    volatilities = tf.convert_to_tensor(
+        volatilities, dtype=dtype, name='volatilities')
+    expiries = tf.convert_to_tensor(expiries, dtype=dtype, name='expiries')
+
+    if discount_rates is not None:
+      discount_rates = tf.convert_to_tensor(
+          discount_rates, dtype=dtype, name='discount_rates')
+      discount_factors = tf.exp(-discount_rates * expiries)
+    elif discount_factors is not None:
+      discount_factors = tf.convert_to_tensor(
+          discount_factors, dtype=dtype, name='discount_factors')
+      discount_rates = -tf.math.log(discount_factors) / expiries
+    else:
+      discount_rates = tf.convert_to_tensor(
+          0.0, dtype=dtype, name='discount_rates')
+      discount_factors = tf.convert_to_tensor(
+          1.0, dtype=dtype, name='discount_factors')
+
+    if dividend_rates is None:
+      dividend_rates = tf.convert_to_tensor(
+          0.0, dtype=dtype, name='dividend_rates')
+
+    if cost_of_carries is not None:
+      cost_of_carries = tf.convert_to_tensor(
+          cost_of_carries, dtype=dtype, name='cost_of_carries')
+    else:
+      cost_of_carries = discount_rates - dividend_rates
+
+    if forwards is not None:
+      forwards = tf.convert_to_tensor(forwards, dtype=dtype, name='forwards')
+    else:
+      spots = tf.convert_to_tensor(spots, dtype=dtype, name='spots')
+      forwards = spots * tf.exp(cost_of_carries * expiries)
+
+    if is_call_options is not None:
+      is_call_options = tf.convert_to_tensor(is_call_options, dtype=tf.bool,
+                                             name='is_call_options')
+    else:
+      is_call_options = tf.constant(True, name='is_call_options')
+
+    if modified_boundary is not None:
+      modified_boundary = tf.convert_to_tensor(modified_boundary, dtype=tf.bool,
+                                             name='modified_boundary')
+    else:
+      modified_boundary = tf.constant(True, name='modified_boundary')
```

Comment:
modified_boundary should just be a boolean. Depending on its value, we build a different graph, so no need to convert this to a tensor. 

This is similar to when we generate random/quasi-random numbers. See, e.g., [here](https://github.com/google/tf-quant-finance/blob/5f47cea348a5d19bd07167cdcdda703118ae0c16/tf_quant_finance/math/random_ops/multivariate_normal.py#L192).

Reply:
Also, please raise NotImplementedError when modified_boundary is True

Reply:
Let me know if the way I implemented this is the way you were expecting.